pyproject.toml

```diff
@@ -65,7 +65,8 @@ linters = "pycodestyle,pyflakes,pylint,mccabe,mypy,radon"
 # R1702:    Too many nested blocks
 # E0401:    Cannot find implementation or library stub for module named
 # R0902:    Too many instance attributes
-ignore = "E402,C901,C0103,E203,R0912,R0915,R901,R901,R0914,C0413,C0206,R1716,W1514,R1702,E0401,R0902"
+# W0511:    TODOs or FIXMEs
+ignore = "E402,C901,C0103,E203,R0912,R0915,R901,R901,R0914,C0413,C0206,R1716,W1514,R1702,E0401,R0902,W0511"
```

Comment:
Do we have documentation/justification for adding `W0511`?

Adding this to your local project is fine but I think it would be best-practice to make sure we're not pushing/keeping `TODO`'s in mainline.

Reply:
I'm going to go with it's perverse incentive. I personally use TODOs and FIXMEs for things things that are worth noting, but not so much that they're worth writing a ticket over, and also not specific to the current commit/pull request I'm working on, or are enough effort that they'd be good in their own commit.

I agree that having TODOs and FIXMEs accumulate is not the best practice, they should be attended to eventually, but I think failing CI on them is worse. I was hoping that I could convince `pylama` to emit a warning and not have it count as a fail, but I didn't see how to convince it to do so.

The perverse incentive is that if this fails CI, I will continue to write these notes but now not label them with TODO or FIXME, making them harder to find. Or just not return to the feature. Both those options seem worse.

Reply:
I'd argue that you can and should be able to continue to use `TODO`'s and all the other useful labels- so long as they are in a stub branch off-mainline and that anything with `TODO`-attached code should be bisected out of the to-merge branch before merging the to-merge branch into main.

I get where you're coming from, I just think there's a way to keep those labels for dev and not prod- is all.

Happy to hear other thoughts and opinions too.

@ryanpdx  what do you think of a *"no todo's in main"*-esqe policy for prod repos?

Reply:
Please do not use `TODO` or `FIXME`, they lead to technical debt

Reply:
I think it's fine to use them- so long as they're all resolved before merge.

If they aren't resolved- then why? Is it part of a feature that should be its own PR? Is it a task item that should be cleaned up first? etc.

Reply:
> Please do not use `TODO` or `FIXME`, they lead to technical debt

They are a symptom not a cause, right? Cart-before-the-horse and all that. Technical debt is caused by lack of $resource during development. A lack of time, of knowledge, of design and planning, of stable requirements, and so on. Some debt is known at the time of writing but can't or won't be addressed due to the aforementioned $resource shortage. Why would we not want that explicitly called out in the code? The author is best positioned to identify these known instances, why would we want to discourage them from telling others about that? The tech debt will be there whether they comment on it or not.

Good comments answer the question why. Why is the code written the way that it is? What was going through the author's mind, what was their intent when writing this piece of code? `TODO`s and `FIXME`s can fit in that category because they answer the question "why not?". The author didn't check this invariant, or implement this in the way I would have expected, or left out this seemingly obvious part of a feature because they lacked $resource, not because it was an intentional design choice. I as someone visiting the codebase later can have more confidence in the changes I make knowing I'm helping the author instead of accidentally stepping on their toes. 

Code comments explicitly are a good way of marking this because of locality. As the codebase evolves, since the comments are right there, they have the best chance of getting updated and staying relevant. If the note lives far away from the code it will drift more. It's not obvious from the code itself that there might be a ticket attached to that specific line, or that there's a branch off from 245 commits back that references it.

Adding the annotation makes the reviewer's job easier too by letting the areas that specifically need more review stand out. Reviewers can help resolve them if they find the issue easier (that they have access to $resource that the submitter did not), or just generally be kept aware of what to look out for in the future.

> I think it's fine to use them- so long as they're all resolved before merge.
> 
> If they aren't resolved- then why? Is it part of a feature that should be its own PR? Is it a task item that should be cleaned up first? etc.

This is effectively the current policy, CI is only run on pushes or merge requests to master, not on branches. The challenge is that code with `TODO`s attached to them is still useful, it accomplishes the task set out for the PR sufficiently, but it hasn't quite lived up to the author's vision. I agree that asking why they weren't resolved is worthwhile, and reviewers can help solve them, but sometimes the answer genuinely is some issues can't or shouldn't be resolved now. For example: waiting on an upstream library to release a new version to use a feature from it. There's nothing the author can do about that but wait. (if you're meticulous you can spot at least one instance of this in this very PR, but mea culpa I didn't mark it).

Code goes to die on stale feature branches. It's hard enough convincing people to read code on master, not a single person will read the stub branch besides the author.  Stale branches can drift quickly from master without active maintenance, merges can get hairy quickly. When revisiting the issue, now not only are there the concerns of the issue itself, there's the extra barrier of fighting through a merge on top. On master the comments get to transform with the code, it's always merged and up to date.

Another benefit of `TODO`s is they're good first project candidates. A new codebase is intimidating, and these provide clear points of study for people looking to contribute. New people are typically only going to look through master.

Reply:
Well, I disagree with everything you just said. I have dealt with too many codebases full of `FIXME`s, `TODO`s, and commented out code where everyone is unsure if they are still relevant. 

But we have no coding standards and this PR is not the place to discuss/decide on them, so I will approve it as everything looks good.